flink-streaming-java/src/main/java/org/apache/flink/streaming/util/HDFSCopyFromLocal.java

```diff
@@ -62,4 +64,34 @@ public void run() {
 			throw asyncException.f0;
 		}
 	}
+
+	/**
+	 * Ensure that target path terminates with a new directory to be created by fs. If remoteURI does not specify a new
+	 * directory, append local directory name.
+	 * @param fs
+	 * @param localPath
+	 * @param remoteURI
+	 * @return
+	 * @throws IOException
+	 */
+	protected static URI checkInitialDirectory(final FileSystem fs,final File localPath, final URI remoteURI) throws IOException {
+		if (localPath.isDirectory()) {
+			Path remotePath = new Path(remoteURI);
+			if (fs.exists(remotePath)) {
+				return new Path(remotePath,localPath.getName()).toUri();
+			}
+		}
+		return remoteURI;
+	}
+
+	protected static void copyFromLocalFile(final FileSystem fs, final File localPath, final URI remotePath) throws Exception {
```

Comment:
Ufuk, I'm a bit confused about this. flink-core FileSystem does not implement copyFromLocal, and both callers of this are using the native hadoop FileSystem. It would be great not to copy code, but it appears all hadoop dependencies are abstracted out of flink-core, so it would not be there.